Fix issues when generating [java] wrappers without package prefix

hoxovic opened the pull request:
When no package is specified in an lcm file, the package from the previously processed file will "leak" to the file currently being processed.

It's considered bad practice not to specify a package, of course, but this fix makes the CMake script more robust in situations where you either cannot just add a package (e.g. inter-company collaboration with shared lcm files) or when you're working with legacy code that you don't want to update.

Thanks

mwoehlke-kitware commented on lcm-cmake/lcmUtilities.cmake:
```diff
@@ -270,6 +270,8 @@ function(lcm_wrap_types)
   foreach(_lcmtype ${_UNPARSED_ARGUMENTS})
     set(_package "")
     set(_outputs "")
+    unset(_package_dir)
```
I think `set(X "")` would be more consistent.

hoxovic replied:
That's a fair point.
Updated

mwoehlke-kitware commented on lcm-cmake/lcmUtilities.cmake:
```diff
@@ -329,7 +331,10 @@ function(lcm_wrap_types)
+          if(_package_dir STREQUAL "")
+            set(_package_dir "lcmtypes")
+          endif()
+            _lcm_add_outputs(_JAVA_SOURCES ${_package_dir}/${_type}.java)
```
BTW, accidental change of indent?

hoxovic replied:
The reason for implementing it for Java only is because I wanted the behaviour to be consistent with `lcm-gen`. There's no C# support in `lcmUtilities.cmake` and as for go (lcm 1.4) I get _Not yet implemented: go-default-package_. That leaves only Java.

I can split it into two PR's if you wish. What do you prefer?

Oh, the indentation was accidentally left behind after making some final changes. I'll fix that.

mwoehlke-kitware replied:
Okay, I guess I can buy that. I'd feel better if someone more familiar with LCM can also weigh in, though. (@ashuang maybe?) I mostly know the CMake bits, and that's about it :wink:.

BTW, if the issue is that certain languages require a package name, maybe it would make sense to move this logic up to line 328? And add a comment e.g. "Following languages require a package name". Having it inside a check for a language being enabled might lead to unexpected behavior, if it is intended to apply to multiple languages.

(I'm not familiar with Go, so don't know what's going wrong there...)

hoxovic replied:
Oops, I saw this comment now.

The default package for C# is different from java's default package, so there needs to be language checks anyway.
I guess it's cleaner to move them out as you suggest, but with only java being able to use a default package at the moment (unless that's changed since 1.4; can't try it since I'm on my phone), the refactoring would basically mean adding an additional check for java and not much else. Shouldn't be too bad to wait with the refactoring till C# wrapper generation is implemented in `lcmUtilities.lcm` and/or `lcm-gen` uses the default package switch for Go, I suppose.

mwoehlke-kitware replied:
Ack. Well, the problem is, if they're different, then the current logic will have changed the package for anything downstream of Java. (Which is nothing, but that can't be true for all languages.) So I think instead you would need to have a language-specific package name that is used, or else tweak the logic to know to unset the package name again after calling `_lcm_add_outputs`. Possibly something as simple as:
```cmake
if(_package_dir STREQUAL "")
  _lcm_add_outputs(_JAVA_SOURCES lcmtypes/${_type}.java)
else()
  _lcm_add_outputs(_JAVA_SOURCES ${_package_dir}/${_type}.java)
endif()
```
Oh, and... I just realized, I think `_package_dir` should be reset to `"."`, not `""`? Otherwise it looks like you end up with absolute paths where you don't want them. (This *definitely* makes me think the 'no package specified' path is poorly tested :slightly_smiling_face:.) Note that this means changing the above `if`, also.

hoxovic replied:
I fully agree with your comment!
Funny thing is, that is what I had before I tried to make it less verbose/repetitive 🙂. Instead I ended up with a prettier but more naïve implementation that could introduce inter-language leaks instead.

I see that the changes have been approved, but I think it's best I update it.

Also, resetting `_package_dir` to `"."` instead isn't a bad idea. I suppose it won't make any functional difference since `_DESTINATION` is set to `${CMAKE_CURRENT_BINARY_DIR}` if it's not defined, and if I'm not mistaken, all paths are prepended with `_DESTINATION` (or `_GO_DESTINATION`) before they are used.
Nevertheless, resetting to `"."` seems more robust and will also make sure there aren't any paths containing consecutive slashes stored in `_CPP_HEADERS`.

hoxovic commented:
I've squashed the commits and also changed the commit message to mention that the default package name is used for java when no package is specified.
I think it'd be better to rename the PR to "Fix `lcmUtilities.cmake` crashing when generating java wrappers without package prefix" or similar rather than having two PR's since that's what I really meant to fix with this PR.

mwoehlke-kitware commented:
> I think it'd be better to rename the PR [...]

Done. You *can* do that yourself, BTW...

hoxovic commented:
Sorry for force-pushing this twice (it was requested I squash the commits).
As per the discussion above, it's preferred to avoid setting `_package_dir` to the default value for java since different languages use different default package prefixes.
